File: ChanhuiSeok/[3]Programmers/땅따먹기.cpp
#include <iostream>
#include <vector>
using namespace std;

int solution(vector<vector<int> > land)
{
    int answer = 0;
    int N = land.size();

    for (int i = 0; i < N - 1; i++) {
        land[i + 1][0] = land[i + 1][0] + max(land[i][1], max(land[i][2], land[i][3]));
        land[i + 1][1] = land[i + 1][1] + max(land[i][0], max(land[i][2], land[i][3]));
        land[i + 1][2] = land[i + 1][2] + max(land[i][0], max(land[i][1], land[i][3]));
        land[i + 1][3] = land[i + 1][3] + max(land[i][0], max(land[i][1], land[i][2]));
    }

    answer = max(land[N - 1][0], max(land[N - 1][1], max(land[N - 1][2], land[N - 1][3])));

    return answer;
}

File: soongjamm/백준/땅따먹기.java
public class 땅따먹기 {

    int solution(int[][] land) {
        int answer = 0;
        int[][] memory = new int[land.length][4];

        for (int row = 0; row < land.length; row++) {
            for (int past = 0; past < 4; past++) {
                if (row == 0) {
                    memory[0][past] = land[0][past];
                    continue;
                }

                for (int col = 0; col < 4; col++) {
                    if (past == col) {
                        continue;
                    }
                    memory[row][col] = Math.max(memory[row - 1][past] + land[row][col], memory[row][col]);
                }
            }
        }

        for (int i = 0; i < 4; i++) {
            answer = Math.answer(answer, memory[land.length - 1][i]);
        }
        return answer;
    }

}